### PR TITLE
Allow postgresql to cross compile

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
###### Motivation for this change

Currently, `PostgreSQL` does not cross compile due to not listing `libXML2` headers as a native build dependency properly, and `libossp-uuid` as a dependency of `PostgreSQL` does not cross compile because it does not respect `strip` command setting made by `configure`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/libossp-uuid/default.nix [hunk 1]
```diff
@@ -12,6 +12,12 @@ stdenv.mkDerivation {
 
   configureFlags = stdenv.lib.optional stdenv.isFreeBSD "--with-pic";
 
+  patches = [ ./shtool.patch ];
+
+  preConfigure = ''
+  export ac_cv_va_copy=yes
```
Comment (dezgeg) [inline pkgs/development/libraries/libossp-uuid/default.nix]: I think conventionally in Nixpkgs, this sort of stuff is not exported but added to `configureFlags` instead.

File: pkgs/servers/sql/postgresql/default.nix [hunk 2]
```diff
@@ -14,8 +14,15 @@ let
     outputs = [ "out" "lib" "doc" "man" ];
     setOutputFlags = false; # $out retains configureFlags :-/
 
+    combinedLibXML2 = symlinkJoin {
```
Comment (dezgeg) [inline pkgs/servers/sql/postgresql/default.nix]: Hmm, why is this needed?

File: pkgs/development/libraries/libossp-uuid/default.nix (hunk 1 above)
Comment (dingxiangfei2009) [reply pkgs/development/libraries/libossp-uuid/default.nix]: Thats the problem. This `configure` script does not allow you to fix the test results on `va_copy`. Please advise a better approach, if any.

File: pkgs/servers/sql/postgresql/default.nix (hunk 2 above)
Comment (dingxiangfei2009) [reply pkgs/servers/sql/postgresql/default.nix]: Possibly, but I just realised that the upstream broke cross-compilation capability, possibly due to #44068 by @Ericson2314. I will close this PR for now.

**Comment by dingxiangfei2009** [comment]
Cross compilation is broken. Close for now and will come back later.

File: pkgs/servers/sql/postgresql/default.nix (hunk 2 above)
Comment (dingxiangfei2009) [reply pkgs/servers/sql/postgresql/default.nix]: @dezgeg I tried. However, the `configure` of `postgresql` expects the `libxml2` path to contain both its headers and its static libraries.

We would either patch the `configure`, or just supply the combined `libxml2`.

Comment (Ericson2314) [reply pkgs/servers/sql/postgresql/default.nix]: If you pass it as a `buildInput`, its dev outputs will be passed with the headers, and the `-L` to the libs added by cc-wrapper will probably leave the configure script unable to notice the difference. If you make it just a nativeBuildInput you will not get the `-L`.

Comment (dingxiangfei2009) [reply pkgs/servers/sql/postgresql/default.nix]: @Ericson2314 I just confirmed that if `libxml2` is added to `buildInputs`, the static library is visible to the cross compilers. However, the headers are not picked up by the `configure` script.

Sorry that I am new to this cross compiling business. Should we modify `CFLAGS` to point to the headers?

Comment (dingxiangfei2009) [reply pkgs/servers/sql/postgresql/default.nix]: Currently `libxml2` export headers in the `include/libxml2` folder in the Nix store, but the `CFLAGS` is pointing to `include` only. When `configure` is resolving `#include <libxml/parser.h>`, it is looking at the `include` folder only and this fails the `configure` checks.

The build works if I point the headers to `include/libxml2`. The question now is whether `libxml2` Nix expression evaluating correctly.

File: pkgs/servers/sql/postgresql/default.nix [hunk 3]
```diff
@@ -22,14 +22,18 @@ let
 
     makeFlags = [ "world" ];
 
+    preConfigure = ''
+    export CFLAGS="$CFLAGS -I${libxml2.dev}/include/libxml2"
```
Comment (matthewbauer) [inline pkgs/servers/sql/postgresql/default.nix]: I think we usually prefer to do this like:

```
NIX_CFLAGS_COMPILE = "-I${libxml2.dev}/include/libxml2";
```

Mostly this is because not all build systems recognize CFLAGS.

File: pkgs/servers/sql/postgresql/default.nix [hunk 4]
```diff
@@ -22,14 +22,18 @@ let
 
     makeFlags = [ "world" ];
 
+    preConfigure = ''
+    export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -I${libxml2.dev}/include/libxml2"
```
Comment (Mic92) [inline pkgs/servers/sql/postgresql/default.nix]: Does postgresql also picks this up, when `pkgconfig` is included in `nativeBuildInputs`?

Comment (Mic92) [reply pkgs/servers/sql/postgresql/default.nix]: Now I see why this fails: 

  https://github.com/postgres/postgres/blob/master/configure.in#L896

It needs `xml2-config` compiled for the build platform.

Comment (Mic92) [reply pkgs/servers/sql/postgresql/default.nix]: Instead of preConfigure, this can be exported in nix instead:

```nix
# needed for cross-compiling, since xml2-config is not compiled for the build platform
NIX_CFLAGS_COMPILE = [ "-I${libxml2.dev}/include/libxml2" ];
```

Comment (Mic92) [reply pkgs/servers/sql/postgresql/default.nix]: The long-term solution would be `pkg-config`: However this would been an upstream fix: https://github.com/libarchive/libarchive/issues/407

Comment (dingxiangfei2009) [reply pkgs/servers/sql/postgresql/default.nix]: I see. I will try to raise this issue in `postgresql` mailing list.

Comment (dezgeg) [reply pkgs/servers/sql/postgresql/default.nix]: Actually `xml2-config` is built correctly but it doesn't get into the `PATH` since we don't add binaries from `dev` output to `PATH`. This is something we should do in stdenv since many other packages will benefit from that as well.

Comment (dingxiangfei2009) [reply pkgs/servers/sql/postgresql/default.nix]: Should I do it in a separate pull request? Or this one?

**Review by matthewbauer — APPROVED** [review]
Great work!

File: pkgs/servers/sql/postgresql/default.nix (hunk 4 above)
Comment (dezgeg) [reply pkgs/servers/sql/postgresql/default.nix]: Yes I agree it can be done later.

Let's however conditionalize it on cross, so that someone doesn't break it by noticing it native-compiles fine with it.

**Comment by Ericson2314** [comment]
Let's fix those few nits first though.

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build pkgsCross.raspberryPi.postgresql

File: pkgs/development/libraries/libossp-uuid/default.nix [hunk 5]
```diff
@@ -12,6 +12,12 @@ stdenv.mkDerivation {
 
   configureFlags = stdenv.lib.optional stdenv.isFreeBSD "--with-pic";
 
+  patches = [ ./shtool.patch ];
+
+  configureFlags = [
+    "ac_cv_va_copy=yes"
```
Comment (matthewbauer) [inline pkgs/development/libraries/libossp-uuid/default.nix]: @ericson2314 Configure flags is also above

**Comment by Ericson2314** [comment]
@GrahamcOfBorg build pkgsCross.raspberryPi.postgresql retroactively...

**Comment by Ericson2314** [comment]
retroactively testing oops (I missed that above, but I also figured at least native builds would continue to work).

**Comment by Ericson2314** [comment]
Cool :)

**Comment by FRidh** [comment]
@Ericson2314 this was a mass-rebuild and should have gone to staging.

**Comment by Ericson2314** [comment]
Sorry @FRidh. I missed that completely.
